File: packages/ts/jns42-generator-ts/src/utils/schema.ts
export function discoverSchemaId(node: unknown) {
  if (node == null) {
    return null;
  }

  if (typeof node !== "object") {
    return null;
  }

  if (!("$schema" in node)) {
    return null;
  }

  if (typeof node.$schema !== "string") {
    return null;
  }

  return node.$schema;
}

File: packages/cargo/jns42-generator/src/utils/schema.rs
use serde_json::Value;
use url::Url;

use crate::documents::meta::MetaSchemaId;

pub fn discover_schema_uri(node: &Value) -> Option<MetaSchemaId> {
    match node {
        Value::Object(object_value) => match object_value.get("$schema") {
            Some(Value::String(string_value)) => Some(string_value.as_str().into()),
            _ => None,
        },
        _ => None,
    }
}
